### PR TITLE
Add timing functionality

### LINKED ISSUE #291: Add option to show timings for each command in cli and also return timings when used programmatically
It would be useful if the package tracked the start and end times of processes and/or the calculated duration in ms, then returning this as part of `ExitInfos` for programmatic use or printing to terminal when a timing option is enabled.

### PR DESCRIPTION
**Pull request by eliasm307**
Add timing functionality to resolve #291

### CODE REVIEW

File: src/completion-listener.js [hunk 1]
```diff
@@ -2,9 +2,10 @@ const Rx = require('rxjs');
 const { bufferCount, switchMap, take } = require('rxjs/operators');
 
 module.exports = class CompletionListener {
-    constructor({ successCondition, scheduler }) {
+    constructor({ successCondition, scheduler, showTimings}) {
```
Comment (gustavohenke) [inline src/completion-listener.js:5]: Why was this needed?

File: src/flow-control/log-timings.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+module.exports = class LogTimings extends BaseHandler {
+    printExitInfoTimingTable(exitInfos) {
+        const exitInfoTable = _(exitInfos)
+            .map(({ command, timings, index, killed, exitCode }) => {
+                const readableDurationMs = (timings.endDate - timings.startDate).toLocaleString();
+                return {
+                    'call-index': index,
+                    name: command.name,
+                    duration: `~${readableDurationMs}ms`,
```
Comment (gustavohenke) [inline src/flow-control/log-timings.js:14]: Why the `~`? Is there any rounding somewhere I'm not spotting?

File: src/flow-control/log-timings.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+const { bufferCount, take } = require('rxjs/operators');
+const BaseHandler = require('./base-handler');
+
+module.exports = class LogTimings extends BaseHandler {
+    printExitInfoTimingTable(exitInfos) {
+        const exitInfoTable = _(exitInfos)
+            .map(({ command, timings, index, killed, exitCode }) => {
+                const readableDurationMs = (timings.endDate - timings.startDate).toLocaleString();
+                return {
+                    'call-index': index,
```
Comment (gustavohenke) [inline src/flow-control/log-timings.js:12]: Is this meant to be different from `(index)`?

```
$ concurrently --show-timings "echo hi"
[0] echo hi started at 11/18/2021, 10:12:25 PM
[0] hi
[0] echo hi stopped at 11/18/2021, 10:12:25 PM after 18ms
[0] echo hi exited with code 0

Timings:
┌─────────┬────────────┬──────┬──────────┬───────────┬────────┬───────────┐
│ (index) │ call-index │ name │ duration │ exit-code │ killed │  command  │
├─────────┼────────────┼──────┼──────────┼───────────┼────────┼───────────┤
│    0    │     0      │  ''  │ '~18ms'  │     0     │ false  │ 'echo hi' │
└─────────┴────────────┴──────┴──────────┴───────────┴────────┴───────────┘
```

File: src/flow-control/log-timings.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                    name: command.name,
+                    duration: `~${readableDurationMs}ms`,
+                    'exit-code': exitCode,
```
Comment (gustavohenke) [inline src/flow-control/log-timings.js:15]: For some slightly added niceness 😊 

```suggestion
                    'exit code': exitCode,
```

File: src/flow-control/log-timings.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                    killed,
+                    command: command.command,
+                };
+            })
+            .sortBy('duration')
```
Comment (gustavohenke) [inline src/flow-control/log-timings.js:20]: Interesting, why sort by duration?

File: bin/concurrently.js [hunk 6]
```diff
@@ -62,6 +62,11 @@ const args = yargs
             default: defaults.hide,
             type: 'string'
         },
+        'show-timings': {
```
Comment (gustavohenke) [inline bin/concurrently.js:65]: What do you think if this was only `--timings`?

File: src/flow-control/log-timings.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    handle(commands) {
+        if (!this.logger) { return {commands}; }
+
+        // individual process timings
+        commands.forEach(command => {
+            command.timer.subscribe( {
+                next: ({startDate, endDate}) => {
+                    if (!endDate) {
+                        this.logger.logCommandEvent( `${ command.command } started at ${ startDate.toLocaleString() }`, command );
```
Comment (gustavohenke) [inline src/flow-control/log-timings.js:36]: Wondering if we should use the format specified in `--timestamp-format` instead (hint, date-fns is a dependency!).

Might make _part_ of the test simpler/more stable on diverse machines due to all those regular expressions...

**Review by gustavohenke — COMMENTED** [review]
whoaa this is super nice! I've added a bunch of thoughts, let me know what you think please.

There are some styling inconsistencies (but these aren't defined anywhere!), so I'll fix them up when we've agreed on the discussion below.

File: src/flow-control/log-timings.js (hunk 2 above)
Comment (eliasm307) [reply src/flow-control/log-timings.js:14]: its just because the duration returned from in the exit info is more accurate, down to the nanosecond, and the two durations might not match but I guess adding this tilde just adds noise to the logs, I'll remove it

File: src/completion-listener.js (hunk 1 above)
Comment (eliasm307) [reply src/completion-listener.js:5]: its a leftover from an old idea, will remove

File: src/flow-control/log-timings.js (hunk 3 above)
Comment (eliasm307) [reply src/flow-control/log-timings.js:12]: so the `(index)` is the index in the sorted array, `call-index` is the index in terms of the order the processes were started, do you have an idea for a better name to make this clear? the `(index)` column is auto generated so cant change that easily unfortunately

File: src/flow-control/log-timings.js (hunk 5 above)
Comment (eliasm307) [reply src/flow-control/log-timings.js:20]: I think if users are using the timing functionality they would be interested in the duration rankings ie knowing what was the fastest, slowest etc at a glance. Duration used to be a number but it wont sort properly as a string, I'll fix this

File: bin/concurrently.js (hunk 6 above)
Comment (eliasm307) [reply bin/concurrently.js:65]: sounds good, I'll update

File: src/flow-control/log-timings.js (hunk 7 above)
Comment (eliasm307) [reply src/flow-control/log-timings.js:36]: good idea, I'll look into it

File: src/flow-control/log-timings.js (hunk 3 above)
Comment (eliasm307) [reply src/flow-control/log-timings.js:12]: maybe it doesnt matter here actually, I'll remove the call index and just leave the sorted index

**Comment by eliasm307** [comment]
@gustavohenke I've applied your feedback, let me know if you see anything else

**Comment by eliasm307** [comment]
@gustavohenke the test coverage being almost 100% was annoying me so I've added a few tests to cover the remaining lines

**Review by gustavohenke — APPROVED** [review]
Thanks! LGTM, I'm just implementing a logger-friendly `console.table` and will push directly to your branch.
We need to respect a bunch of other combinations users can have (raw mode, using a custom stdout, etc)
